### ISSUE TITLE
More intuitive error messages for configuration file problems

### ISSUE THREAD

**Issue by ericmjonas** [nchammas/flintrock#65]
I'm just returning to using flintrock, and I recloned and edited a new config.yaml, and recived several errors along the way. 

### Identity file check
the identity file field does not recognize the unix homedir convention of `~/.ssh/foo.bar.baz` but will still launch the cluster, and the result can be confusing. Can we explicitly test for the existence of the file? 

### region / ami mismatch
by default you might want to change a region, but it's not obvious you have to also change your ami and when you run you receive this message: 

```
boto.exception.EC2ResponseError: EC2ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>InvalidAMIID.NotFound</Code><Message>The image id '[ami-61bbf104]' does not exist</Message></Error></Errors><RequestID>f27df620-5903-4403-b1e1-77309b265e86</RequestID></Response>
```

This is a little bit more helpful, but still a bit confusing. 

### username / ami mismatch
Finally, if you use the wrong user with your ami everything works until flintrock tries to ssh in and you get this confusing paramiko error: 

```
Exception: Error reading SSH protocol banner[Errno 54] Connection reset by peer
Traceback (most recent call last):
  File "/Users/jonas/anaconda/envs/py3k/lib/python3.5/site-packages/paramiko/transport.py", line 1724, in _check_banner
    buf = self.packetizer.readline(timeout)
  File "/Users/jonas/anaconda/envs/py3k/lib/python3.5/site-packages/paramiko/packet.py", line 326, in readline
    buf += self._read_timeout(timeout)
  File "/Users/jonas/anaconda/envs/py3k/lib/python3.5/site-packages/paramiko/packet.py", line 479, in _read_timeout
    x = self.__socket.recv(128)
ConnectionResetError: [Errno 54] Connection reset by peer
```

**Comment by nchammas** [comment]
* Identity file: We do check for the existence of the file [here](https://github.com/nchammas/flintrock/blob/ee27fc20a0d2615dbb4e5f5ae753019505c0d797/flintrock.py#L458-L459). Perhaps you used an incorrect identity file, as opposed to a non-existent identity file?
* Region - AMI mismatch: This is trickier to address, but it's covered by #24, I believe. One thing we could look into as an interim solution is whether you can query the AMI's metadata for what region it belongs to and error out early if there is a mismatch.
* Username - AMI mismatch: This is impossible to address upfront unless, again, there is a way to query the AMI's metadata for the user. Do you know if that's possible? Otherwise, the only other approach here would be to offer some kind of `--resume` functionality like spark-ec2, but I'd rather not add that for now.

**Comment by nchammas** [comment]
Yeah, I just checked what we do for bad identity files, and Flintrock already detects this cleanly:

```sh
$ flintrock launch nick
Usage: flintrock launch [OPTIONS] CLUSTER_NAME

Error: Invalid value for "--ec2-identity-file": Path "~/.ssh/nick.pem" does not exist.
```
```
$ flintrock launch nick
Usage: flintrock launch [OPTIONS] CLUSTER_NAME

Error: Invalid value for "--ec2-identity-file": Path "/Users/nicholaschammas/.ssh/bad.pem" does not exist.
```

Are you not getting this behavior?

By the way, Flintrock cannot handle `~/` because that is a feature of your shell (e.g. Bash) and not the filesystem. The shell automatically expands `~/` for you before passing it on as arguments to your command-line programs. But if you put this in a config file, then your shell is not doing this expansion, and whatever program reads that config file gets a path it does not understand.

I'm working on friendlier error messages for the other two cases you described.

**Comment by nchammas** [comment]
Regarding bad usernames, this is going to be tough to address.

The exception we get back from Paramiko is `paramiko.ssh_exception.SSHException: Error reading SSH protocol banner`. This can be caused by [a variety of things](https://www.google.com/search?q=Error+reading+SSH+protocol+banner) and I've hit it myself in the past for different reasons.

If there is a way to tell more specifically that the username might be the issue, I'm all for adding a hint to the user to check that. Without something more to go on, adding a hint may just create more confusion, especially if the real issue is something other than the username.

The only other thing I can offer here is to clean up how we handle exceptions in general as part of cluster launches (#48), since you do get a lot of traceback barf from the various running threads.

So in summary, for the 3 issues in your original message:

1. I believe this is already well addressed.
2. PR coming to add a friendlier message here.
3. Hard to do anything specific to bad usernames, but some cleanup coming as part of #48.
